### PR TITLE
use dune in the Makefile and remove ocamlbuild

### PR DESCRIPTION
**Pull request by zapashcanon**
Hi,

As discussed in #1664, I switched everything from `ocamlbuild` to `dune`, I removed winmake.bat along the way. I made a lot of simplifications in the Makefile but also removed a few things: we don't build the bytecode version anymore and we don't copy the executable at the root anymore.

### CODE REVIEW

File: interpreter/README.md
Comment (ngzhian) [inline interpreter/README.md:1]: I think we should add instructions here on how to get `dune` https://github.com/WebAssembly/spec/tree/main/interpreter#building

File: interpreter/Makefile [hunk 1]
```diff
@@ -9,131 +9,39 @@
...
-DIRS =		util syntax binary text valid runtime exec script host main tests
-LIBS =		
-FLAGS = 	-lexflags -ml -cflags '-w +a-4-27-42-44-45-70 -warn-error +a-3'
-OCBA =		ocamlbuild $(FLAGS) $(DIRS:%=-I %)
-OCB =		$(OCBA) $(LIBS:%=-libs %)
-JSO =		js_of_ocaml -q --opt 3
-JS =		# set to JS shell command to run JS tests, empty to skip
+OPT =		wasm.exe
+ZIP =		wasm.zip
+JSLIB =		wast.bc.js
```
Comment (rossberg) [inline interpreter/Makefile:14]: The end result here should still be named wast.js. May require renaming it in the build target.

File: interpreter/README.md [hunk 2]
```diff
@@ -22,10 +22,7 @@ Once you have OCaml, simply do
 ```
 make
 ```
-You'll get an executable named `./wasm`. This is a byte code executable. If you want a (faster) native code executable, do
-```
-make opt
-```
+You'll get an executable named `_build/default/wasm.exe`.
```
Comment (rossberg) [inline interpreter/README.md:25]: Let's not break people's workflow. Can you tweak the makefile such that this is symlinked to ./wasm?

File: interpreter/Makefile [hunk 3]
```diff
@@ -9,131 +9,39 @@
 
 # Configuration
 
-NAME =		wasm
```
Comment (rossberg) [inline interpreter/Makefile:12]: Can we keep the NAME variable? That makes it easier to reuse the Makefile for derivatives.

File: interpreter/Makefile [hunk 4]
```diff
@@ -182,27 +83,8 @@ dunetest:
...
-		git archive --format=zip --prefix=$(NAME)/ -o $@ HEAD
+$(ZIP):
+	git archive --format=zip --prefix=wasm/ -o $@ HEAD
 
 clean:
-		rm -rf _build/jslib $(LIB).mlpack _tags $(JSLIB).js
-		$(OCB) -clean
-
-
-# Opam support
```
Comment (rossberg) [inline interpreter/Makefile:193]: Why did you remove the opam support?

File: interpreter/README.md
Comment (rossberg) [reply interpreter/README.md:1]: Agreed.

File: interpreter/Makefile [hunk 5]
```diff
@@ -9,131 +9,39 @@
...
-		$(OCB) -quiet $@
-
-.PHONY:		smallint.byte smallint.native
-smallint.byte: _tags
-		$(OCB) -quiet $@
-smallint.native: _tags
-		$(OCB) -quiet $@
-
-
-# Building library
```
Comment (rossberg) [inline interpreter/Makefile:82]: This is used to create the Wasm opam package. Perhaps it can be simplified, but we still need it.

**Review by rossberg — COMMENTED** [review]
Changes look good mostly, but you may have been a bit over-aggressive in deleting stuff. At least I wouldn't know how to build the opam library package anymore after this. ;)

Is there a way to avoid and work around the renaming main.ml→wasm.ml? IMO it's bad design that dune assumes this.

File: interpreter/Makefile (hunk 4 above)
Comment (zapashcanon) [reply interpreter/Makefile:193]: It's broken currently. I'll need to make a few fixes to get it back (e.g. move the `dune-project` at the root of the repository again, see [ocaml/dune#7962](https://github.com/ocaml/dune/issues/7962). Currently it crashes, when it'll be fixed, we'll still get an error saying that dependencies outside of workspace is not supported. You can reproduce easily by doing `dune runtest` from the `interpreter` directory on current main.

File: interpreter/README.md (hunk 2 above)
Comment (zapashcanon) [reply interpreter/README.md:25]: I'm fine with having `./wasm` but IIRC, currently it's `./wasm.exe` and we get an error from dune saying that we're asking to generate `./wasm.exe` which is already in the source tree.

File: interpreter/Makefile (hunk 5 above)
Comment (zapashcanon) [reply interpreter/Makefile:82]: If you're OK with it, I can bring back proper opam support in another PR, but it requires more fixes (e.g. having a `public_name` and not simply a `name` etc.).

File: interpreter/Makefile (hunk 3 above)
Comment (zapashcanon) [reply interpreter/Makefile:12]: Fixed.

File: interpreter/Makefile (hunk 4 above)
Comment (zapashcanon) [reply interpreter/Makefile:193]: I made a minimal opam support.

**Comment by zapashcanon** [comment]
> Is there a way to avoid and work around the renaming main.ml→wasm.ml? IMO it's bad design that dune assumes this.

I did something for this. Everything should be OK now @rossberg. :)

File: interpreter/Makefile [hunk 6]
```diff
@@ -142,67 +55,45 @@ TESTDIR =	../test/core
...
+	dune install
 
 # Miscellaneous targets
 
 .PHONY:		clean
 
-$(ZIP):		$(WINMAKE)
-		git archive --format=zip --prefix=$(NAME)/ -o $@ HEAD
+$(ZIP):
+	git archive --format=zip --prefix=wasm/ -o $@ HEAD
```
Comment (rossberg) [inline interpreter/Makefile:95]:
```suggestion
	git archive --format=zip --prefix=$(NAME)/ -o $@ HEAD
```

File: interpreter/Makefile [hunk 7]
```diff
@@ -142,67 +55,45 @@ TESTDIR =	../test/core
...
-
-install:	_build/$(LIB).cmx _build/$(LIB).cmo
-		ocamlfind install $(LIB) meta/findlib/META _build/$(LIB).o \
-		  $(wildcard _build/$(LIB).cm*) \
-		  $(wildcard $(DIRS:%=%/*.mli))
-
-uninstall:
-		ocamlfind remove $(LIB)
+	dune clean
+	rm -f $(NAME) $(JSLIB)
```
Comment (rossberg) [inline interpreter/Makefile:99]: This should probably be in a separate distclean target.

File: interpreter/Makefile [hunk 8]
```diff
@@ -9,131 +9,44 @@
...
-		echo if \'%1\' neq \'\' set NAME=%1 >>$@
-		$(OCB) main.byte \
-		| grep -v ocamldep \
-		| grep -v mkdir \
-		| sed s:`which ocaml`:ocaml:g \
-		| sed s:main/main.d.byte:%NAME%.exe: \
-		>>$@
-
+$(JSLIB):
+	rm -f $@
```
Comment (rossberg) [inline interpreter/Makefile:47]: That is odd, why is this necessary? You probably want to make the copy via a target, so that dependency tracking works properly.

File: interpreter/wasm.opam [hunk 9]
```diff
@@ -0,0 +1,22 @@
+# This file is generated by dune, edit dune-project instead
```
Comment (rossberg) [inline interpreter/wasm.opam:1]: If this file is generated, is there a reason to check it in?

File: interpreter/.gitignore [hunk 10]
```diff
@@ -1,13 +1,3 @@
-*.cmo
-*.cmx
-*.native
-*.byte
-*.opt
-*.unopt
-*.js
-*.zip
```
Comment (rossberg) [inline interpreter/.gitignore:8]: You probably want to keep *.js and *.zip

**Review by rossberg — COMMENTED** [review]
Cool, thanks!

File: interpreter/Makefile (hunk 6 above)
Comment (zapashcanon) [reply interpreter/Makefile:95]: Fixed.

File: interpreter/Makefile (hunk 7 above)
Comment (zapashcanon) [reply interpreter/Makefile:99]: Fixed.

File: interpreter/wasm.opam (hunk 9 above)
Comment (zapashcanon) [reply interpreter/wasm.opam:1]: Yes, that's standard. The main reason is that, you usually want to be able to run things like: `opam pin git@github.com:WebAssembly/spec.git`. But opam doesn't know that the package is to be built with `dune` and can't call `dune build wasm.opam`, thus we have to commit the file. There are plans to fix this but I don't know if it'll happen soon.

File: interpreter/.gitignore (hunk 10 above)
Comment (zapashcanon) [reply interpreter/.gitignore:8]: Fixed.

File: interpreter/Makefile [hunk 11]
```diff
@@ -92,8 +92,10 @@ install:
 .PHONY:		clean
 
 $(ZIP):
-	git archive --format=zip --prefix=wasm/ -o $@ HEAD
+	git archive --format=zip --prefix=$(NAME)/ -o $@ HEAD
 
 clean:
 	dune clean
+
+distclean:
```
Comment (rossberg) [inline interpreter/Makefile:100]:
```suggestion
distclean: clean
```

**Review by rossberg — APPROVED** [review]

File: interpreter/Makefile (hunk 11 above)
Comment (zapashcanon) [reply interpreter/Makefile]: Sorry, fixed.

File: interpreter/wasm.opam (hunk 9 above)
Comment (avsm) [reply interpreter/wasm.opam:1]: It's not a bug, so no plans to fix it :-)

Checking in the `opam` file ensures that the package manager can just inspect that file from the release archive to know what to do with it.  If code needs to be run to generate the file, and that fails or otherwise has complex constraints (e.g. a different dune), then figuring out an installation plan becomes quite complex.

So this looks fine the way it is...

File: interpreter/Makefile (hunk 8 above)
Comment (rossberg) [reply interpreter/Makefile:47]: This should work:
```
$(JSLIB):	$(BUILDDIR)/$(JSLIB)
	cp $< $@

$(BUILDDIR)/$(JSLIB):
	dune build $(JSLIB)
```

File: interpreter/Makefile [hunk 12]
```diff
@@ -142,67 +55,47 @@ TESTDIR =	../test/core
...
-# Opam support
-
-.PHONY:		check install uninstall
-
-check:
-		# Check that we can find all relevant libraries
-		# when using ocamlfind
-		ocamlfind query $(LIBS)
-
-install:	_build/$(LIB).cmx _build/$(LIB).cmo
```
Comment (rossberg) [inline interpreter/Makefile:202]: This target (as well as opt and unopt) is currently used by the `opam` package file. I'm not sure what that needs to be replaced with.

As part of a doing a new 2.0.1 package this morning, I checked in a template for the opam file under meta/opam. Can you merge main and adapt that file to the changes here?

File: interpreter/wasm.opam (hunk 9 above)
Comment (rossberg) [reply interpreter/wasm.opam:1]: I see. Somehow you folks must have a higher itch threshold against cluttered directories than I... ;)

**Review by rossberg — COMMENTED** [review]
Sorry, I had somehow lost this PR.

File: interpreter/wasm.opam (hunk 9 above)
Comment (zapashcanon) [reply interpreter/wasm.opam:1]: > we might even skip the generation step altogether by having opam invoke dune to get the generated opam file directly

From [discuss]. But I guess the plans changed one more time. ;-)

File: interpreter/Makefile (hunk 12 above)
Comment (zapashcanon) [reply interpreter/Makefile:202]: They won't be needed anymore once we use dune.

I updated the `dune-project` to add the new metadata from `meta/opam`.

**Comment by zapashcanon** [comment]
It should be good now @rossberg ?

File: interpreter/Makefile (hunk 8 above)
Comment (rossberg) [reply interpreter/Makefile:47]: But what does that have to do with the Makefile? Wouldn't you get that error either way?

Comment (zapashcanon) [reply interpreter/Makefile]: The Makefile is the one putting `wast.js` at the root of the repository. It works with your fix as long as one doesn't try to run `make wast.js` and then `dune build wast.js`.

I'm not going to use the Makefile anyway so I don't really care. And I believe people using the Makefile are not going to run dune manually so this is probably fine...

File: interpreter/Makefile [hunk 13]
```diff
@@ -142,78 +55,47 @@ TESTDIR =	../test/core
...
-opam-release/%:
-		git tag opam-$*
-		git push --tags
-		rm -f opam-$*.zip
-		wget https://github.com/WebAssembly/spec/archive/opam-$*.zip
-		cp meta/opam/opam .
-		sed -i ".tmp" s/@VERSION/$*/g opam
-		sed -i ".tmp" s/@MD5/`md5 -q opam-$*.zip`/g opam
-		rm opam.tmp opam-$*.zip
-		@echo Created file ./opam, submit to github opam-repository/packages/wasm/wasm.$*/opam
```
[the remark this answers is not in the document]
Comment (rossberg) [reply interpreter/Makefile:219]: Thanks for volunteering, I am happy to pass that on to you. :)

Yet, others should still be able to do that easily in case, to avoid single points of failure. A makefile target is a useful abstraction that is both automating and documenting the steps necessary, and tends to be a more stable interface than the tools behind it. (Once a person has worked on a dozen diverse systems that each have gone through multiple generations of custom build tools, they start to appreciate having that abstraction. ;) )

If opam-publish or dune-release are the best way of implementing that process now (I admit I haven't used those before), can you adjust the Makefile target to call into those?

File: interpreter/Makefile [hunk 14]
```diff
@@ -88,6 +88,13 @@ install:
 	dune build -p $(NAME) @install
 	dune install
 
+.PHONY: opam-release
```
Comment (rossberg) [inline interpreter/Makefile:91]: I don't think this works, since the target contains a wildcard.

**Review by rossberg — APPROVED** [review]
Thanks!

File: interpreter/Makefile (hunk 14 above)
Comment (zapashcanon) [reply interpreter/Makefile:91]: Oh yes sorry, it's because at first I had a `opam-release:` target without a Makefile.

File: interpreter/Makefile [hunk 15]
```diff
@@ -88,6 +88,13 @@ install:
...
+
+opam-release/%:
+	git tag opam-$*
+	git push --tags
+	opam publish
```
[the remark this answers is not in the document]
Comment (zapashcanon) [reply interpreter/Makefile:94]: IIRC it's going to ask.

[the remark this answers is not in the document]
Comment (rossberg) [reply interpreter/Makefile:94]: I just tried opam-publish with all variations and available options, but it always said "Not enough information to publish", even when I set the version explicitly. My suspicion is that it bakes in assumptions like package-name = repo-name.

The implementation details don't really matter to me, but this target should continue working.

**Review by rossberg — CHANGES_REQUESTED** [review]

File: interpreter/Makefile (hunk 15 above)
[the remark this answers is not in the document]
Comment (zapashcanon) [reply interpreter/Makefile:94]: Indeed, it breaks also `dune-release`. This is because we don't have `wasm.opam` at the root of the repository...
Should I move `dune-project` and `wasm.opam` at the root (I know you don't like it) ? Otherwise we can keep the previous version.

[the remark this answers is not in the document]
Comment (zapashcanon) [reply interpreter/Makefile:94]: Also note that having them at the root of the repository will make `opam pin https://github.com/WebAssembly/spec` to work properly.

[the remark this answers is not in the document]
Comment (rossberg) [reply interpreter/Makefile:94]: By root you mean the root of the repo? I'm afraid that wouldn't be adequate, since the interpreter is only a secondary subcomponent of this project that shouldn't affect the root dir. It would also slightly break path-dependent CI.

Would perhaps opams-publish's --packages-directory=interpreter option work in this case, by any chance?

[the remark this answers is not in the document]
Comment (rossberg) [reply interpreter/Makefile:94]: That's unfortunate, but thanks for poking. Let's revert to what was there before then. Maybe opam will support it at some point in the future.

(CI is set up per toplevel directory, i.e., interpreter builds only run when something under interpreter/ changes. Moving dune and opam files out of there would then fail to capture them. Although that could probably be hacked.)

File: interpreter/Makefile [hunk 16]
```diff
@@ -91,7 +91,17 @@ install:
...
+	rm -f opam-$*.zip
+	wget https://github.com/WebAssembly/spec/archive/opam-$*.zip
+	cp wasm.opam opam
+	echo 'url {' >> opam
+	echo '  src: "https://github.com/WebAssembly/spec/archive/opam-@VERSION.zip"' >> opam
+	echo '  checksum: "md5=@MD5"' >> opam
+	echo '}' >> opam
+	sed -i ".tmp" s/@VERSION/$*/g opam
+	sed -i ".tmp" s/@MD5/`md5 -q opam-$*.zip`/g opam
+	rm opam.tmp opam-$*.zip
```
Comment (rossberg) [inline interpreter/Makefile:103]: If we are producing this via echo, we don't need sed but can splice in the right strings right here:
```suggestion
	echo "  src: \"https://github.com/WebAssembly/spec/archive/opam-$*.zip\"" >> opam
	echo "  checksum: \"md5=`md5 -q opam-$*.zip`\"" >> opam
	echo "}" >> opam
        rm opam-$*.zip
```

**Review by rossberg — APPROVED** [review]

File: interpreter/Makefile (hunk 16 above)
Comment (zapashcanon) [reply interpreter/Makefile:103]: Indeed. Also the `sed -i ".tmp" ...` part is not working on linux, this is a macOS thing I guess. To be portable you'd do something like `sed -i.tmp ...`.

File: interpreter/Makefile [hunk 17]
```diff
@@ -95,12 +95,10 @@ opam-release/%:
 	wget https://github.com/WebAssembly/spec/archive/opam-$*.zip
 	cp wasm.opam opam
 	echo 'url {' >> opam
```
Comment (rossberg) [inline interpreter/Makefile:97]: Nit: maybe also use `"` here for consistency.

**Review by rossberg — APPROVED** [review]

File: interpreter/Makefile (hunk 17 above)
Comment (zapashcanon) [reply interpreter/Makefile]: done :)

**Comment by rossberg** [comment]
I merged it, thanks for taking this.
